lib/runner.js

```diff
@@ -724,42 +729,22 @@ Runner.prototype.uncaught = function (err) {
     return;
   }
 
-  runnable.clearTimeout();
-
   // Ignore errors if already failed or pending
   // See #3226
   if (runnable.isFailed() || runnable.isPending()) {
     return;
   }
   // we cannot recover gracefully if a Runnable has already passed
   // then fails asynchronously
-  var alreadyPassed = runnable.isPassed();
-  // this will change the state to "failed" regardless of the current value
-  this.fail(runnable, err);
-  if (!alreadyPassed) {
-    // recover from test
-    if (runnable.type === 'test') {
-      this.emit('test end', runnable);
-      this.hookUp('afterEach', this.next);
-      return;
-    }
+  if (runnable.isPassed() || !runnable.callback) {
```

Comment:
I'm unclear what we're doing with the `callback`

Reply:
The callback is what is to be executed if the test fails. It's the `done()` function of the current test.

If there is a current test, then that test should be cancelled and all after hooks should be run on any uncaught exceptions.
If there is no current test, then uncaught exceptions simply mark the current suite as failed.